src/OrbitBase/include/OrbitBase/NotFoundOr.h

```diff
@@ -20,44 +20,64 @@ struct NotFound {
 };
 
 // NotFoundOr can be used as the return type of a search operation, where the search can be
-// unsuccessful and returns a message. Based on std::variant. Check whether a NotFoundOr object is
+// unsuccessful and returns a message. Check whether a NotFoundOr object is
 // a "not found" result with orbit_base::IsNotFound. Get the message with GetNotFoundMessage.
+//
+// `NotFoundOr<T>` is implemented by privately deriving from `std::variant`. This allows us to take
+// advantage of std::variant's carefully crafted constructors while still keeping std::variant an
+// implementation detail. Implementing those constructors by hand would add a lot of complexity.
 template <typename T>
-using NotFoundOr = std::variant<VoidToMonostate_t<T>, NotFound>;
+class NotFoundOr : private std::variant<VoidToMonostate_t<T>, NotFound> {
```

Comment:
`NotFoundOr` and `CanceledOr` have 4 identical methods. I suggest to add a common ancestor like `template <Base> Wrapper : protected Base` that would implement public `HasValue` and `GetValue`s. Then we inherit from `Wrapper` twice.

Reply:
Is there a way to avoid publicly inheriting from the `Wrapper` and still do the same? I really don't wanna introduce a base type that's part of the public API at this point. Like I said I would prefer not to design a generic wrapper class at this point.

Reply:
We can put Wrapper in `orbit_interal` namespace.
Alternatively, we can always inherit privately and then have 4 using clauses in each of the subclasses.

Reply:
... actually, just 2 using clauses, as a using clause would pull all the overloads.

Reply:
I definitely would prefer the private inheritance and the using declarations in that case.

But do you really think that this is worth it? I mean there is really not a lot of code shared. And you need to introduce a new header file with a lot of lines of boilerplate just to avoid 4 duplicated lines, let alone the readability by not having these simple functions in the same file? Also finding a name for this common private base class, writing a comment that explain why this class exists. I struggle to see that this is the right focus.

Reply:
Right now we have `const Value& NotFoundOr::GetValue() const &&` and `const Value&& CanceledOr::GetValue() const&&`. Is this difference intentional? 

Fighting code duplication is not about typing or code length. It's about making the code coherent and making it easy to maintain such coherence under changes.

Reply:
Well but that's the point. This code just happens to be the same. There is no contract in place which requires these two types to be coherent. I can change the function names to `GetNonCanceled` and `GetFound` if that helps.

Not intentional. I can fix that.

Reply:
There is such contract. The one we've got in mind -- we believe if we call `GetValue` on rvalue, rvalueref should be returned. The current implementation of `NotFoundOr` doesn't adhere to it, therefore we agree it should be fixed. 

Neither the initial commit adhered. Due to code duplication the fix is now applied to only one place of the two. What if there are still bugs in the code we can't currently see? Will we have to fix it in two places next time as well?

Reply:
Still, it just happens to be the same.

Reply:
> Still, it just happens to be the same
Does "just happens" mean "not compile-time checked"? 

1. There was a bug in the code
2. The bug wouldn't be there if the code was de-duplicated

Isn't it enough of an argument?

Reply:
`ValueOr`

Reply:
... or `Either`. Because it resembles the Either monad.